Add messageExpression field for CRD validation

DangerOnTheRanger opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature
/kind api-change

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

This PR implements the KEP update outlined in https://github.com/kubernetes/enhancements/pull/3747.

#### Special notes for your reviewer:

The planned API changes have been implemented in this PR (so discussion/API review can start), but this PR currently pins a dependency on https://github.com/google/cel-go/pull/645. When that PR merges and a `cel-go` version bump PR is merged into k8s, then the WIP tag will be taken off. Other than the pin, the PR should be considered ready to merge and reviewed as such.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Added messageExpression field to ValidationRule. (#115969, @DangerOnTheRanger)
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/pull/3747
```

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -251,6 +252,26 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
+				} else {
+					messageStr, ok := evalResult.Value().(string)
+					if !ok {
+						klog.Warningf("could not could message expression result to string, falling back: %v", evalResult.Value())
```
The message: `could not resolve messageExpression to string`?

cici37 replied:
And stop here if `!ok`? We don't want to append err if cannot resolve to string right?

DangerOnTheRanger replied:
Right - I've re-worked the error paths (and fixed the typo in the message, terribly sorry about that).

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -251,6 +252,26 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
+			if len(rule.MessageExpression) != 0 {
+				evalResult, evalDetails, err := compiled.MessageExpression.ContextEval(ctx, activation)
+				if evalDetails == nil {
+					klog.Warningf("runtime cost could not be calculated for message expression %q")
```
Also, we still want do runtime cost check on messageExpression right?
And also count the runtime cost of messageExpression towards the whole budget as well.

DangerOnTheRanger replied:
I've added a cost check, but I'll hold off on resolving this till I add a test that verifies the cost check works. You're right with `evalDetails == nil` - I've adjusted it to fall back to `Message` in that case.

DangerOnTheRanger replied:
(marking as resolved now that there's unit tests for the runtime cost check)

cici37 commented on test/integration/apiserver/crd_validation_expressions_test.go:
```diff
@@ -523,8 +523,8 @@ func TestCustomResourceValidatorsWithBlockingErrors(t *testing.T) {
-				if !strings.Contains(err.Error(), "failed rule: self.spec.x + self.spec.y") {
-					t.Fatalf("Expected error to contain %s but got %v", "failed rule: self.spec.x + self.spec.y", err.Error())
+				if !strings.Contains(err.Error(), "self.spec.x + self.spec.y must be greater than or equal to") {
+					t.Fatalf("Expected error to contain %s but got %v", "self.spec.x + self.spec.y must be greater than or equal to", err.Error())
```
Have the full string here for testing? For test case, we should specify what we expected to have from `format([(has(self.status) ? self.status.z : 0)]`

DangerOnTheRanger replied:
Right, I've added the output from the `%d` portion of the `string.format` call to make sure that part works as well too.

DangerOnTheRanger commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/celcoststability_test.go:
```diff
@@ -331,10 +331,10 @@ func TestCelCostStability(t *testing.T) {
 			}),
 			// https://github.com/google/cel-spec/blob/master/doc/langdef.md#field-selection
 			expectCost: map[string]int64{
```
`has()` reports a runtime cost of 0 and not 1 after pinning https://github.com/google/cel-go/pull/645. I thought that might not be expected, so for the time being I isolated the changes in a separate commit.

I agree that it's a regression though - I won't mark this as resolved until I find the root cause and fix it.

jpbetz replied:
Would you bisect to find the cel-go commit that introduced the change?

DangerOnTheRanger replied:
(marking as resolved due to opening https://github.com/google/cel-go/pull/651)

DangerOnTheRanger commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types_jsonschema.go:
```diff
@@ -196,6 +196,13 @@ type ValidationRule struct {
+	// If both Message and MessageExpression are present on a Rule, then MessageExpression will be used if validation
+	// fails. If MessageExpression results in a runtime error, then Message will be used instead if non-empty. If
+	// Message is empty, then the default error message is used instead. In either case, the CEL error that MessageExpression
+	// failed with will be logged.
```
`MessageExpression` is just a CEL expression that evaluates to a string and doesn't have to be used with `string.format`  (`self.str1 + "some constant" + self.str2` would be just as valid) - I agree the syntax for `string.format` should be documented somewhere (outside of the existing reference documentation in cel-go), but should we put it here in the k8s API documentation, as opposed to something closer to cel-go? I definitely think some examples would be nice though.

`//+ optional` is also a good point, I'll add that as well.

DangerOnTheRanger replied:
(marking as resolved - `// +optional` is included, `string.format`-specific docs are left out of this PR in preparation for a follow-up)

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types_jsonschema.go:
```diff
@@ -196,6 +196,13 @@ type ValidationRule struct {
 	// e.g. "must be a URL with the host matching spec.host"
 	Message string
+	// MessageExpression represents a CEL expression that can be evaluated when validation fails, whereas Message
+	// is just a static string. MessageExpression must statically evaluate to a string to be accepted.
```
I don't think the majority of users will understand what "statically evaluate" means here. Clarify or use a different term?  Is "MessageExpression must evaluate to a string" sufficient?

DangerOnTheRanger replied:
I revised it to read "must evaluate to a string at compilation time" instead, which seems like it'd be pretty readable.

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types_jsonschema.go:
```diff
@@ -196,6 +196,13 @@ type ValidationRule struct {
 	// If unset, the message is "failed rule: {Rule}".
 	// e.g. "must be a URL with the host matching spec.host"
 	Message string
+	// MessageExpression represents a CEL expression that can be evaluated when validation fails, whereas Message
```
I wouldn't bring up Message here since it is already covered in detail below.

DangerOnTheRanger replied:
Gotcha, I removed that bit.

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types_jsonschema.go:
```diff
@@ -196,6 +196,13 @@ type ValidationRule struct {
 	// If unset, the message is "failed rule: {Rule}".
 	// e.g. "must be a URL with the host matching spec.host"
 	Message string
+	// MessageExpression represents a CEL expression that can be evaluated when validation fails, whereas Message
```
"MessageExpression declares a CEL expression that evaluates to the validation failure message that is returned when this rule fails" ?

DangerOnTheRanger replied:
Right, fixed

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types_jsonschema.go:
```diff
@@ -196,6 +196,13 @@ type ValidationRule struct {
 	Message string
+	// MessageExpression represents a CEL expression that can be evaluated when validation fails, whereas Message
+	// is just a static string. MessageExpression must statically evaluate to a string to be accepted.
+	// If both Message and MessageExpression are present on a Rule, then MessageExpression will be used if validation
+	// fails. If MessageExpression results in a runtime error, then Message will be used instead if non-empty. If
+	// Message is empty, then the default error message is used instead. In either case, the CEL error that MessageExpression
+	// failed with will be logged.
```
Can we simplify the conditional logic in the text here? I found it a bit twisty to read.

DangerOnTheRanger replied:
Would you prefer it as a list, maybe something like:

```markdown
If MessageExpression results in a runtime error, then it will be logged, and the fallback message will instead use:

* Message (if non-empty)
* The default error message (if Message is empty)
```

I left "If both Message and MessageExpression are present on a Rule, then MessageExpression will be used if validation fails" out of the edit since I thought the hard part to understand was with the runtime error resolution part, but let me know if that's not the case.

jpbetz replied:
"If MessageExpression results in a runtime error, the validation failure message is produced as-if the MessageExpression field were unset" ?

DangerOnTheRanger replied:
Gotcha, sounds good to me. Fixed.

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go:
```diff
@@ -206,6 +209,32 @@ func compileRule(rule apiextensions.ValidationRule, env *cel.Env, perCallLimit u
+	if rule.MessageExpression != "" {
+		ast, issues := env.Compile(rule.MessageExpression)
+		if issues != nil {
+			compilationResult.Error = &apiservercel.Error{apiservercel.ErrorTypeInvalid, "messageExpression compilation failed: " + issues.String()}
```
Why is the MessageExpression compilation error being used to populate the compliationResult.Error for the main expression?

DangerOnTheRanger replied:
Yeah, I think that's confusing. I've split `MessageExpression`-related errors out into a separate field.

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go:
```diff
@@ -206,6 +209,32 @@ func compileRule(rule apiextensions.ValidationRule, env *cel.Env, perCallLimit u
+			return
+		}
+		if ast.OutputType() != cel.StringType {
+			compilationResult.Error = &apiservercel.Error{apiservercel.ErrorTypeInvalid, "messageExpression must evaluate to a string"}
```
same

DangerOnTheRanger replied:
Also fixed

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go:
```diff
@@ -206,6 +209,32 @@ func compileRule(rule apiextensions.ValidationRule, env *cel.Env, perCallLimit u
+
+		_, err := cel.AstToCheckedExpr(ast)
+		if err != nil {
+			compilationResult.Error = &apiservercel.Error{apiservercel.ErrorTypeInternal, "unexpected messageExpression compilation error: " + err.Error()}
```
same

DangerOnTheRanger replied:
Also fixed

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation_test.go:
```diff
@@ -641,6 +641,63 @@ func TestCelCompilation(t *testing.T) {
 				invalidError("must evaluate to a bool"),
 			},
 		},
+		{
```
Why no parse error, incorrect return type and type check error cases?

DangerOnTheRanger replied:
Right, I've added compilation tests that check for parse errors/non-string return types/`string.format` issues.

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation_test.go:
```diff
@@ -660,6 +717,7 @@ func TestCelCompilation(t *testing.T) {
 
 			for i, expectedResult := range tt.expectedResults {
 				if !expectedResult.matches(compilationResults[i]) {
+					t.Logf(compilationResults[i].Error.Error())
```
Handle error checking here properly with `t.Errorf` or `t.Fatalf`?

DangerOnTheRanger replied:
This line was a little fragile + hard to generalize with the addition of `MessageExpressionError`, so I've removed it.

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -251,6 +252,26 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
+					klog.Warningf("runtime cost could not be calculated for message expression %q")
+				}
+				if err != nil {
+					klog.Warningf("error evaluating message expression, falling back: %v", err.Error())
```
We should keep an eye on this. I'm a bit concerned someone will flood logs with this.

jpbetz replied:
An alternative would be to return a client warning..

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -251,6 +252,26 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
+					tryMessage = false
+				}
+			}
+			if !tryMessage {
+				continue
+			}
```
Is `tryMessage` needed? Is it better than invoking`continue` at the same place where `tryMessage` is set to `false`?

DangerOnTheRanger replied:
I've refactored the function a bit, so now `tryMessage` is gone.

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -251,6 +252,26 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
 			continue
 		}
 		if evalResult != types.True {
+			tryMessage := true
```
This function is getting out of hand. Can we split some of this message related logic into a utility?

DangerOnTheRanger replied:
Right, I've added `evalMessageExpression`.

cici37 commented:
/triage accepted

cici37 commented:
For easy to review purpose and better history tracking, could you please organize your commits as following when ready for review/merge?
1. cel-go dependency bump(it could be in a separate PR and merge separately)
2. API manual changes - Add MessageExpression field
3. Auto changes(auto generated files, dependencies, etc)
4. The implementation regarding with the field added
You could always have temp commits for addressing comments and squash them when ready to merge :)
Thank you :)

liggitt commented:
> @liggitt Curious on your thoughts? Thank you

looks like we're hitting https://github.com/kubernetes/kubernetes/issues/113366

if there's any way to revert / avoid that, we should take it

cici37 commented:
> > @liggitt Curious on your thoughts? Thank you
> 
> looks like we're hitting #113366
> 
> > Work with cel-go to update/revert genproto version to fulfill Kubernetes dependency requirement which will be a temporary solution and I could see further update request coming with similar issue.
> 
> if there's any way to revert, we should do that... we gain 100+ additional dependencies in our tree if we take this bump

Thanks for the info!

I guess the timing is great since we haven't cut the cel-go release yet. @DangerOnTheRanger will open the PR to revert genproto version in cel-go and we could discuss there. Thank you

cici37 commented:
After revisit and tried locally, this one is actually not as hug as https://github.com/kubernetes/kubernetes/issues/113366.
It only touched the very low end of the dependency graph and removing the restriction on `cloud.google.com/go/firestore` in  [unwanted-dependencies.json](https://github.com/kubernetes/kubernetes/blob/67f4f9ad5a3d8653b63652774442a09e7780b78b/hack/unwanted-dependencies.json#L4) could easily solve the issue. Since `cloud.google.com/go/firestore` will be removed anyway due to genproto is introducing it, could we lift the restriction and unblock the cel-go bump this way? @liggitt

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types_jsonschema.go:
```diff
@@ -196,6 +196,22 @@ type ValidationRule struct {
 	// If unset, the message is "failed rule: {Rule}".
 	// e.g. "must be a URL with the host matching spec.host"
 	Message string
+	// MessageExpression declares a CEL expression that evaluates to the validation failure message that is returned when this rule fails.
```
"MessageExpression is a CEL expression that is evaluated when the rule fails to produce a failure message for the rule." ?

DangerOnTheRanger replied:
Doesn't that make it sound a bit like `MessageExpression` is a fallback when `Message` or the default message don't work for some reason?

jpbetz replied:
Yes, that doesn't always parse right. You're fix of putting the "when this rule fails" at the end of the phrase is the correct. fix.

liggitt commented:
> Since the restriction on cloud.google.com/go/firestore will be removed eventually, could we lift the restriction and unblock the cel-go bump this way?

There's no intention of removing that restriction... we explicitly do not need or want a dependency on a database client

DangerOnTheRanger commented:
I force-pushed to remove code related to `string.format` while we work on getting the firestore issue worked out with cel-go - we'll land anything related to `string.format` in a separate PR.

cici37 reviewed (commented):
Test coverage could be improved. We should at least have tests for:
1. Combination of all possible exists status for `messageExpression`, `message`, `rule`; 
2. `messageExpression` fail to compile
3. `messageExpression` fail to eval to string
4. err handling on different level:  on compilation, validation, API validation, etc.
5. runtime cost exceed for `messageExpression`
6. overall budget exceed for `messageExpression`
7. If the cost calculated correctly for `messageExpression`
8. Should we do static cost check on `messageExpression`?

You might be able to come up with more. :)

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types_jsonschema.go:
```diff
@@ -235,6 +235,21 @@ type ValidationRule struct {
+	// If both Message and MessageExpression are present on a Rule, then MessageExpression will be used if validation
+	// fails. If MessageExpression results in a runtime error, the runtime error is logged, and the validation failure message is produced
+	// as if the MessageExpression field were unset.
+	// If string.format is used inside MessageExpression, then MessageExpression can also fail due to the number of
+	// arguments given to format() not matching the amount requested inside the format string, a type mismatch between
+	// an argument and a formatting clause (%s/%f/%d/etc.), and invalid syntax inside the formatting string itself. In any of those cases,
```
Should all references to `string.format()` be dropped from this PR and added in the PR where `string.format()` is added? Doing so would keep the main branch in a consistent state.

cc @cici37

DangerOnTheRanger replied:
Yeah, I didn't think about how the main branch would refer to non-existent functionality for a time - I'm in favor of removing it as well.

cici37 replied:
+1 to keep it consistent

DangerOnTheRanger replied:
Gotcha, fixed.

DangerOnTheRanger commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -251,16 +254,57 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
-			if len(rule.Message) != 0 {
-				errs = append(errs, field.Invalid(fldPath, sts.Type, rule.Message))
+			if len(rule.MessageExpression) != 0 {
+				messageExpression, newRemainingBudget, err := evalMessageExpression(ctx, compiled.MessageExpression, rule.MessageExpression, activation, remainingBudget)
```
Right, I'll add that and a test to match.

DangerOnTheRanger commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types_jsonschema.go:
```diff
@@ -196,6 +196,21 @@ type ValidationRule struct {
+	// If string.format is used inside MessageExpression, then MessageExpression can also fail due to the number of
+	// arguments given to format() not matching the amount requested inside the format string, a type mismatch between
+	// an argument and a formatting clause (%s/%f/%d/etc.), and invalid syntax inside the formatting string itself. In any of those cases,
+	// the same fallback strategy of using first Message and then the default error message will be tried, in addition
```
(resolving since `string.format` docs are removed from this PR)

DangerOnTheRanger commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -251,16 +254,57 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
+					klog.Warning(err)
+					errs = append(errs, field.Invalid(fldPath, sts.Type, ruleMessageOrDefault(rule)))
+				}
+				errs = append(errs, field.Invalid(fldPath, sts.Type, messageExpression))
```
Right, fixed

DangerOnTheRanger commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -251,16 +254,57 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
+		return "", -1, fmt.Errorf("runtime cost could not be calculated for message expression, falling back: %q", exprSrc)
+	} else {
+		if *rtCost > math.MaxInt64 || int64(*rtCost) > remainingBudget {
+			return "", -1, fmt.Errorf("messageExpression evaluation failed due to running out of cost budget, falling back: %q", exprSrc)
```
I've added a new `messageExpressionErr` type that wraps `error` , and then I check for the runtime error from `validateExpressions` - think this resolves it, feel free to comment further if you have more thoughts though.

cici37 commented:
> 
> On the other hand, we're able to statically guarantee/check that `messageExpression` returns a string here, so maybe that's not that big of an issue.

We do compilation check even in API validation level so we should know if `messageExpression` successfully compiled at this point.

>
> do you think it'd be surprising to users of messageExpression that that wouldn't/couldn't be done for ValidatingAdmissionPolicy though?

It is different since we know exactly the schema in CRD scenario but don't know at ValidatingAdmissionPolicy

Although, I am ok with only applying runtime cost limit to `messageExpression` since we don't really expect a expression successful come in with a much more costly `messageExpression`. :)

DangerOnTheRanger commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -251,16 +254,57 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
-			if len(rule.Message) != 0 {
-				errs = append(errs, field.Invalid(fldPath, sts.Type, rule.Message))
+			if len(rule.MessageExpression) != 0 {
+				messageExpression, newRemainingBudget, err := evalMessageExpression(ctx, compiled.MessageExpression, rule.MessageExpression, activation, remainingBudget)
```
(marking as resolved)

DangerOnTheRanger commented:
/label api-review

cici37 commented:
Note: more tests are expected in [k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/](http://k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation_test.go)[validation_test.go](http://k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation_test.go). Currently no tests added here. Since this is where the API level validation check happens and we have compilation running, it would be a good place to check the user facing error message logic.

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -251,16 +254,89 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
+	if evalDetails == nil {
+		return "", -1, &messageExpressionErr{
+			runtimeCostErr,
+			fmt.Errorf("runtime cost could not be calculated for message expression, falling back: %q", exprSrc),
```
`"runtime cost could not be calculated for message expression: %v", exprSrc`?

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -251,16 +254,89 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
+	if rtCost == nil {
+		return "", -1, &messageExpressionErr{
+			runtimeCostErr,
+			fmt.Errorf("runtime cost could not be calculated for message expression, falling back: %q", exprSrc),
```
`fmt.Errorf("runtime cost could not be calculated for message expression: %v", exprSrc),`?

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -251,16 +254,89 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
+		if *rtCost > math.MaxInt64 || int64(*rtCost) > remainingBudget {
+			return "", -1, &messageExpressionErr{
+				runtimeCostErr,
+				fmt.Errorf("messageExpression evaluation failed due to running out of cost budget, falling back: %q", exprSrc),
```
`messageExpression evaluation failed due to running out of cost budget, no further validation rules will be run`

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -251,16 +254,89 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
+	if !ok {
+		return "", -1, &messageExpressionErr{
+			evalErr,
+			errors.New("could not convert message expression result to string, falling back"),
```
`message expression failed to convert to string`

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -251,16 +254,89 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
+	return messageStr, remainingBudget - int64(*rtCost), nil
+}
+
+func ruleMessageOrDefault(rule apiextensions.ValidationRule) string {
```
Why not directly using `ruleErrorString`?

DangerOnTheRanger replied:
`ruleMessageOrDefault` preserves the current behavior of printing `failed rule: ...` if we don't have `Message`. There are two cases where we want to have that logic:

*  `MessageExpression` results in an error besides exhausting the runtime cost budget
*  `MessageExpression` is not included at all (the user didn't specify one, etc.)

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -251,16 +254,89 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
+				}
 			} else {
-				errs = append(errs, field.Invalid(fldPath, sts.Type, fmt.Sprintf("failed rule: %s", ruleErrorString(rule))))
+				errs = append(errs, field.Invalid(fldPath, sts.Type, ruleMessageOrDefault(rule)))
```
RuleErrorString here already handles using message or rule as returned message back to user right?

DangerOnTheRanger replied:
I think this is the same case as [here](https://github.com/kubernetes/kubernetes/pull/115969#discussion_r1126804627). We could remove `ruleMessageOrDefault`, but I think we'd have to duplicate the logic here twice: https://github.com/kubernetes/kubernetes/blob/0270fc75d0f59e1f5a31a810737ac35ac3eeb8a3/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go#L254-L258

cici37 replied:
I see. This makes sense for having `failed rule:` only when returning rule as message. :)

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -251,16 +254,89 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
-			if len(rule.Message) != 0 {
-				errs = append(errs, field.Invalid(fldPath, sts.Type, rule.Message))
+			if compiled.MessageExpression != nil {
+				messageExpression, newRemainingBudget, msgErr := evalMessageExpression(ctx, compiled.MessageExpression, rule.MessageExpression, activation, remainingBudget)
```
This is tricky here. `runtimeCostErr` you defined actually includes two types of runtimeCostErr: 
- cost failed to retrieve. 
- cost exceeds budge. 

However, the `evalErr` you defined also includes two types of err: 
- cost per expression failed. 
- eval failed.  

In current code, both `cost per expression failed.` and `eval failed` will lead to warning and using message/rule. I guess we would like to keep the behavior same as existing one which is `cost per expression failed` will stop running following expression and return err.

DangerOnTheRanger replied:
Right, the problem is using `continue` and not an immediate `return`. I'll move it to be in line with current behavior and add some tests that check for runtime cost failures.

DangerOnTheRanger replied:
(marking as resolved - cost per expression and eval failure should be properly tested now)

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -251,16 +254,89 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
+						errs = append(errs, field.Invalid(fldPath, sts.Type, ruleMessageOrDefault(rule)))
+					}
+				} else {
+					errs = append(errs, field.Invalid(fldPath, sts.Type, messageExpression))
```
Why append err here?

DangerOnTheRanger replied:
This is the `else` for `if msgErr != nil` - so this means `MessageExpression` evaluated successfully, but we still have a rule invalidation that we need to report. Similar to what we do here: https://github.com/kubernetes/kubernetes/blob/0270fc75d0f59e1f5a31a810737ac35ac3eeb8a3/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go#L253-L256

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -251,16 +254,89 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
+						continue
+					} else {
+						klog.Warning(msgErr.err)
+						errs = append(errs, field.Invalid(fldPath, sts.Type, ruleMessageOrDefault(rule)))
```
reuse `ruleErrorString`?

DangerOnTheRanger replied:
I think this is also the same case as https://github.com/kubernetes/kubernetes/pull/115969#discussion_r1126809763.

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -251,16 +254,89 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
-				errs = append(errs, field.Invalid(fldPath, sts.Type, rule.Message))
+			if compiled.MessageExpression != nil {
+				messageExpression, newRemainingBudget, msgErr := evalMessageExpression(ctx, compiled.MessageExpression, rule.MessageExpression, activation, remainingBudget)
+				if msgErr != nil {
```
The remainingBudget is not handled here?

DangerOnTheRanger replied:
Hmm, you're right. I'll add a test to ensure runtime cost is calculated correctly when a `MessageExpression` evaluation error occurs.

DangerOnTheRanger replied:
(marking as resolved due to https://github.com/kubernetes/kubernetes/pull/115969/commits/f534fc82836446c97210a0faffc906bc9ec81c54)

cici37 commented:
I haven't looked at the tests close yet. But if the existing tests didn't detect https://github.com/kubernetes/kubernetes/pull/115969#discussion_r1125758318, https://github.com/kubernetes/kubernetes/pull/115969#discussion_r1125758364, https://github.com/kubernetes/kubernetes/pull/115969#discussion_r1125760148, there must be tests missing.

cici37 commented:
Regarding with the CI failure, feel free to take this commit: https://github.com/DangerOnTheRanger/kubernetes/pull/3 for openapi spec updates.

cici37 commented:
Note: If https://github.com/kubernetes/kubernetes/pull/116054 merged first, `messageExpression` has to omit `authorizer`.

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -251,16 +254,90 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
+	if len(rule.Message) == 0 {
+		return fmt.Sprintf("failed rule: %s", ruleErrorString(rule))
+	} else {
+		return rule.Message
```
Should here be `strings.TrimSpace(rule.Message)` as what we did following?

DangerOnTheRanger replied:
You're right - fixed

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -251,16 +254,90 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
+	if err != nil {
+		return "", remainingBudget - int64(*rtCost), &messageExpressionErr{
+			evalErr,
+			fmt.Errorf("error evaluating message expression, falling back: %v", err.Error()),
```
nit: keep the err message consistent
```suggestion
fmt.Errorf("message expression evaluation failed due to : %v", err.Error()),
```

DangerOnTheRanger replied:
Gotcha, fixed

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -251,16 +254,90 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
+			}
+		}
+	}
+	if err != nil {
```
Differentiate the cost exceed err from other evalErr?

DangerOnTheRanger replied:
If the cost is exceeded we'll get a `runtimeCostErr`:

```golang
if *rtCost > math.MaxInt64 || int64(*rtCost) > remainingBudget {
	return "", -1, &messageExpressionErr{
		runtimeCostErr,
		errors.New("messageExpression evaluation failed due to running out of cost budget, no further validation rules will be run"),
	}
}
```

That's different from `evalErr` - did you mean differentiate the runtime cost error from other `runtimeCostErr`s? I feel the current set of `runtimeCostErr`s is a pretty reasonable group, since I don't think we want to continue if either a) `evalDetails` is `nil` b) `evalDetails.ActualCost()` is nil c) we're out of budget.

cici37 replied:
This is cost overall budget exceed. I am talking about the per expression limit exceed^^

cici37 replied:
Ref: https://github.com/kubernetes/kubernetes/blob/a5f7d533b08e05dbf0ce23db4bdd62ad9df64a62/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go#L247-L249

DangerOnTheRanger replied:
Gotcha - marking this as resolved once I add a branch + tests.

DangerOnTheRanger replied:
(marking as resolved due to 8ebb12c6b8efb2f517e09ec3b676b7cfbb0f3181)

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -251,16 +254,90 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
 	return errs, remainingBudget
 }
 
+type _errType int
```
nit: Use idiomatic typed errors here?

```go
type expressionError struct{...}
func (*expressionError) Error() string { 
...
}
```
`evalMessageExpression()` can then return `error` and callers can use `errors.As` ?

jpbetz replied:
Note that you probably only need a typed error for runtime cost errors since you could then return all eval errors as normal errors.

DangerOnTheRanger replied:
Right, makes sense, I'll refactor accordingly.

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -251,16 +254,90 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
+	err     error
+}
+
+// evalMessageExpression takes a context, Program and source code representing the MessageExpression, activation,
+// and remaining, runtime cost budget, returns the evaluated message and the remaining budget, or an error if one
```
nit: Don't repeat information already obvious from the function signature unless supplementing it in a useful way. Examples: "takes a context that, if canceled, may result in an XYZ error", "returns an error if the expression fails to execute in the given budget, the context is canceled, ...=

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types_jsonschema.go:
```diff
@@ -196,6 +196,15 @@ type ValidationRule struct {
+	// If both Message and MessageExpression are present on a Rule, then MessageExpression will be used if validation
+	// fails. If MessageExpression results in a runtime error, the runtime error is logged, and the validation failure message is produced
+	// as if the MessageExpression field were unset.
+	// Examples:
+	// "self.x must be less than " + string(self.max)
```
nit: use Example (singular), or add more examples

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation_test.go:
```diff
@@ -8558,6 +8558,95 @@ func TestValidateCustomResourceDefinitionValidation(t *testing.T) {
+			},
+			expectedErrors: []validationMatch{
+				invalid("spec.validation.openAPIV3Schema.properties[f].x-kubernetes-validations[0].messageExpression"),
+			},
```
I was expecting more tests here along the lines of:
- Test failure where expression fails and the error message is produced by the message expression
- Test failure where expression fails and the error message is produced by the message because the message expression encountered a runtime error
- ...

DangerOnTheRanger replied:
Isn't this `validation_test.go` for testing/validating CRDs? I don't think we do similar tests for `Message` in this file either. I think there's tests like what you're describing in the other `validation_test.go` (in `apiserver/schema/cel`) though.

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go:
```diff
@@ -70,6 +70,12 @@ type CompilationResult struct {
 	// MaxCardinality represents the worse case number of times this validation rule could be invoked if contained under an
 	// unbounded map or list in an OpenAPIv3 schema.
 	MaxCardinality uint64
+	// MessageExpression represents the cel Program that should be evaluated to generate an error message if the rule
+	// fails to validate. If no MessageExpression was given, this will be nil.
+	MessageExpression cel.Program
```
also nil if the MessageExpression failed to compile?

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -1046,6 +1046,9 @@ func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSch
 								allErrs.CELErrors = append(allErrs.CELErrors, field.Invalid(fldPath.Child("x-kubernetes-validations").Index(i).Child("rule"), schema.XValidations[i], cr.Error.Detail))
 							}
 						}
+						if cr.MessageExpressionError != nil {
```
Also add a non-empty check to the line 1004 group of validations to match what exists for the rule field?

DangerOnTheRanger replied:
Gotcha, fixed

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types_jsonschema.go:
```diff
@@ -196,6 +196,15 @@ type ValidationRule struct {
+	// Since MessageExpression is used as a failure message, it must evaluate to a string at compilation time.
+	// If both Message and MessageExpression are present on a Rule, then MessageExpression will be used if validation
+	// fails. If MessageExpression results in a runtime error, the runtime error is logged, and the validation failure message is produced
+	// as if the MessageExpression field were unset.
```
nit: Make it super clear that the message CEL expression has access to all the same variables as the `rule` expression and that the only difference is the return type?

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -251,16 +254,90 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
+			runtimeCostErr,
+			fmt.Errorf("runtime cost could not be calculated for message expression: %q", exprSrc),
+		}
+	} else {
+		if *rtCost > math.MaxInt64 || int64(*rtCost) > remainingBudget {
```
nit: `} else if {` here?

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation_test.go:
```diff
@@ -2259,6 +2260,130 @@ func TestCELMaxRecursionDepth(t *testing.T) {
+			expectedValidationErr: "this is messageExpression",
+		},
+		{
+			name:                    "message used instead of messageExpression in case of syntax errors",
+			message:                 "message not messageExpression",
+			messageExpression:       `"unclosed string`,
```
What case is being tested here? If parse fails, the messageExpression cannot be written to the CRD, right?

DangerOnTheRanger replied:
You're right, we won't get this far - I've removed the test.

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation_test.go:
```diff
@@ -2259,6 +2260,130 @@ func TestCELMaxRecursionDepth(t *testing.T) {
 	}
 }
 
+func TestMessageExpression(t *testing.T) {
```
I was expecting to see a test that produced a runtime error.

DangerOnTheRanger replied:
`runtime cost preserved if messageExpression fails during evaluation` maybe? That one fails during evaluation - was there another sort of runtime error you were looking for?

jpbetz replied:
Ah i missed that. LGTM

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go:
```diff
@@ -209,6 +215,36 @@ func compileRule(rule apiextensions.ValidationRule, env *cel.Env, perCallLimit u
+		}
+		compilationResult.MessageExpression = msgProg
+	}
 	return
```
Are we allowing messageExpressions that exceed the max estimated cost limit? I was expecting to see estimated cost supported here (and tested).

jpbetz commented:
> > On the other hand, we're able to statically guarantee/check that `messageExpression` returns a string here, so maybe that's not that big of an issue.
> 
> We do compilation check even in API validation level so we should know if `messageExpression` successfully compiled at this point.
> 
> > do you think it'd be surprising to users of messageExpression that that wouldn't/couldn't be done for ValidatingAdmissionPolicy though?
> 
> It is different since we know exactly the schema in CRD scenario but don't know at ValidatingAdmissionPolicy
> 
> Although, I am ok with only applying runtime cost limit to `messageExpression` since we don't really expect a expression successful come in with a much more costly `messageExpression`. :)

If we don't support it when we introduce the field in this PR, then it would be a breaking change to add it later.

We've already done a _ton_ of work to implemented estimated cost and if we don't support it for `messageExpression`, it significantly reduces the value of having estimated cost, because someone can get around it by using the messageExpression field.

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go:
```diff
@@ -194,6 +203,42 @@ func compileRule(rule apiextensions.ValidationRule, env *cel.Env, perCallLimit u
+			cel.InterruptCheckFrequency(celconfig.CheckFrequency),
+		)
+		costEst, err := env.EstimateCost(ast, estimator)
+		if err != nil {
+			compilationResult.MessageExpressionError = &apiservercel.Error{Type: apiservercel.ErrorTypeInternal, Detail: "cost estimation failed for messageExpression: " + err.Error()}
+			return
+		}
+		if err != nil {
+			compilationResult.MessageExpressionError = &apiservercel.Error{Type: apiservercel.ErrorTypeInvalid, Detail: "messageExpression instantiation failed: " + err.Error()}
+			return
+		}
```
The second error handling case here appears to be misplaced.

DangerOnTheRanger replied:
Whoops, you're right. Fixed

DangerOnTheRanger commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation_test.go:
```diff
@@ -8647,6 +8647,68 @@ func TestValidateCustomResourceDefinitionValidation(t *testing.T) {
 			},
 		},
+		{
+			name: "x-kubernetes-validations rule invalidated by messageExpression exceeding per-expression cost limit",
```
Gotcha, done

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/types_jsonschema.go:
```diff
@@ -235,6 +235,16 @@ type ValidationRule struct {
 	// e.g. "must be a URL with the host matching spec.host"
 	Message string `json:"message,omitempty" protobuf:"bytes,2,opt,name=message"`
+	// MessageExpression declares a CEL expression that evaluates to the validation failure message that is returned when this rule fails.
+	// Since MessageExpression is used as a failure message, it must evaluate to a string at compilation time.
```
nit: remove `at compilation time` since user should not care about the concept of compilation?

DangerOnTheRanger replied:
Good point, I'll remove that.

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -1014,6 +1015,9 @@ func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSch
 			}
+			if len(rule.MessageExpression) > 0 && len(trimmedMsgExpr) == 0 {
+				allErrs.SchemaErrors = append(allErrs.SchemaErrors, field.Required(fldPath.Child("x-kubernetes-validations").Index(i).Child("messageExpression"), "messageExpression must be non-empty if specified"))
+			}
```
Couple things missed earlier: 
- Should we loose the previous validation on `message`?
```
else if len(rule.Message) > 0 && len(trimmedMsg) == 0 {
				allErrs.SchemaErrors = append(allErrs.SchemaErrors, field.Invalid(fldPath.Child("x-kubernetes-validations").Index(i).Child("message"), rule.Message, "message must be non-empty if specified"))
			} else if hasNewlines(trimmedMsg) {
				allErrs.SchemaErrors = append(allErrs.SchemaErrors, field.Invalid(fldPath.Child("x-kubernetes-validations").Index(i).Child("message"), rule.Message, "message must not contain line breaks"))
			} else if hasNewlines(trimmedRule) && len(trimmedMsg) == 0 {
				allErrs.SchemaErrors = append(allErrs.SchemaErrors, field.Required(fldPath.Child("x-kubernetes-validations").Index(i).Child("message"), "message must be specified if rule contains line breaks"))
```
  
- Should we add check on newlines in MessageExpression?

The first question might be no due to backward compatible concern.

DangerOnTheRanger replied:
Makes sense, I'll add a line break check (I'm not sure `message must be specified if rule contains line breaks` is applicable to `messageExpression`).

cici37 replied:
After offline discussion with @jpbetz , let's leave it as it is now since there is no control here on if cel program gonna return multiple lines string or one line. But to keep the behavior sync and avoid returning large message, we could consider:
- through runtime error when messageExpression returned multiple lines string
- add a length limit on the output(If the resulting message is more than 5k, it is not returned and instead the message is used)

DangerOnTheRanger replied:
Sounds good, I'll add those two instead.

DangerOnTheRanger replied:
(marking as resolved due to beb0f1c4ea5e78718584fe10295bb03de15147cb)

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -252,16 +256,100 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
+				messageExpression, newRemainingBudget, msgErr := evalMessageExpression(ctx, compiled.MessageExpression, rule.MessageExpression, activation, remainingBudget)
+				if msgErr != nil {
+					if errors.Is(msgErr, &runtimeCostError{}) {
+						errs = append(errs, field.InternalError(fldPath, msgErr))
```
We previously treat only `evalDetails == nil` as internalError and the rest would be `Invalid` error. Would we keep it consistent?

DangerOnTheRanger replied:
Right, I've added a new `&internalError` type for these situations - fixed

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -252,16 +256,100 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
+	}
+	trimmedMsgStr := strings.TrimSpace(messageStr)
+	if len(trimmedMsgStr) > maxMessageExpressionLen {
+		return "", remainingBudget - int64(*rtCost), errors.New("message expression generated string beyond allowable length")
```
Let's return the allowed max length back to user?

DangerOnTheRanger replied:
Gotcha, fixed

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -252,16 +256,100 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
+	if len(trimmedMsgStr) > maxMessageExpressionLen {
+		return "", remainingBudget - int64(*rtCost), errors.New("message expression generated string beyond allowable length")
+	} else if hasNewlines(trimmedMsgStr) {
+		return "", remainingBudget - int64(*rtCost), errors.New("message expression generated string with newlines")
```
The err message should be sync as `messageExpression should not contain line breaks`. 
In the error message, should we use field name `messageExpression`?

DangerOnTheRanger replied:
Right, all error messages should use `messageExpression` and not `message expression` now, fixed

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -252,16 +256,116 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
+	maxMessageExpressionLen = 5000
+)
+
+type internalError struct {
```
Note: I feel it does not make much sense to have new error type here. We should be able to find existing err to be reused. But we could fix it in follow-up PR.

DangerOnTheRanger replied:
(marking as resolved, `apiservercel.Error` is used in place of an internal error type now)

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go:
```diff
@@ -55,6 +55,15 @@ type CompilationResult struct {
 	// unbounded map or list in an OpenAPIv3 schema.
 	MaxCardinality uint64
+	// MessageExpression represents the cel Program that should be evaluated to generate an error message if the rule
+	// fails to validate. If no MessageExpression was given, or if thee expression failed to compile, this will be nil.
```
nit: `thee` -> `this`

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -252,16 +256,103 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
+	}
+	rtCost := evalDetails.ActualCost()
+	if rtCost == nil {
+		return "", -1, &cel.Error{
```
Instead of using cel.Error and runtimeCostError, Let's use the errorType to differentiate Internal err and Invalid err. I think the current `runtimeCostError` goes to Invalid error right?

DangerOnTheRanger replied:
(this is showing up as not resolved on my end - may have forgotten to resolve it, but `runtimeCostError` has been removed, resolving)

DangerOnTheRanger committed:
Add MessageExpression field.

Update docs to note that generating line breaks from messageExpression is not allowed.

DangerOnTheRanger committed:
Update codegen/openapi spec.

DangerOnTheRanger committed:
Implement MessageExpression.

cici37 commented:
/lgtm
/approve

jpbetz commented:
/lgtm
All my concerns appear to have been addressed.

jpbetz commented:
/assign @liggitt

cici37 commented:
/assign @liggitt 
It is ready for review when you have time :). Thank you!

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -252,16 +255,102 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
+						errs = append(errs, field.Invalid(fldPath, sts.Type, msgErr.Error()))
+						return errs, -1
+					} else {
+						klog.Warning(msgErr)
```
Switch to structural logging?
klog.V(2).ErrorS(*)

DangerOnTheRanger replied:
Right, I'll mark this as resolved when I push that change.

DangerOnTheRanger replied:
Marking as resolved - I do think it's worth noting that the test outputs some logging info to `stderr` now. If this is something I should fix, let me know/reopen and I'll address it.

valaparthvi commented:
Hi @DangerOnTheRanger. I am from the Bug Triage Shadow team for K8s release v1.27. I see this PR is actively worked upon, but I wanted to make sure it is on track for 1.27 release.
Please note that Code Freeze begins on March 15. Post this date, only critical and release blocker PRs/issues will be accepted to the current release.

DangerOnTheRanger commented:
(force push to address https://github.com/kubernetes/kubernetes/pull/115969#discussion_r1128817584)

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types_jsonschema.go:
```diff
@@ -196,6 +196,16 @@ type ValidationRule struct {
+	// as if the MessageExpression field were unset.
+	// MessageExpression has access to all the same variables as the Rule; the only difference is the return type.
+	// Example:
+	// "self.x must be less than " + string(self.max)
```
nit, but we would encourage people to output real field names, not CEL variable names, like this:

```suggestion
	// "x must be less than max ("+string(self.max)+")"
```

(copy that example to the versioned types as well)

DangerOnTheRanger replied:
Right, I'll adjust the docs accordingly.

DangerOnTheRanger replied:
(marking as resolved, example has been changed)

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/types_jsonschema.go:
```diff
@@ -235,6 +235,16 @@ type ValidationRule struct {
 	// If unset, the message is "failed rule: {Rule}".
 	// e.g. "must be a URL with the host matching spec.host"
 	Message string `json:"message,omitempty" protobuf:"bytes,2,opt,name=message"`
+	// MessageExpression declares a CEL expression that evaluates to the validation failure message that is returned when this rule fails.
+	// Since MessageExpression is used as a failure message, it must evaluate to a string.
+	// If both Message and MessageExpression are present on a Rule, then MessageExpression will be used if validation
+	// fails. If MessageExpression results in a runtime error, the runtime error is logged, and the validation failure message is produced
+	// as if the MessageExpression field were unset.
+	// MessageExpression has access to all the same variables as the Rule; the only difference is the return type.
```
use serialized field names in godoc (`messageExpression`,`message`, `rule`, etc) so it matches what users actually set via the API

DangerOnTheRanger replied:
Gotcha, I'll change that and mark as resolved when I do.

DangerOnTheRanger replied:
(marking as resolved - think it's worth pointing out that I kept the first mention unchanged from `MessageExpression` since this seems consistent with godoc rules + usage in the rest of the file; if I misunderstood let me know and I can fix that)

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/types_jsonschema.go:
```diff
@@ -235,6 +235,16 @@ type ValidationRule struct {
+	// Since MessageExpression is used as a failure message, it must evaluate to a string.
+	// If both Message and MessageExpression are present on a Rule, then MessageExpression will be used if validation
+	// fails. If MessageExpression results in a runtime error, the runtime error is logged, and the validation failure message is produced
+	// as if the MessageExpression field were unset.
```
> If MessageExpression results in a runtime error, the runtime error is logged, and the validation failure message is produced as if the MessageExpression field were unset.

I would also expect that fallback if the expression evaluates to "" or null. is that what happens? if so, document and test for that

DangerOnTheRanger replied:
Thanks for catching this, this wasn't tested. I'll document it and test for it as well.

DangerOnTheRanger replied:
(marking as resolved, docs now note that evaluating to an empty string is not allowed, added tests as well)

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/types_jsonschema.go:
```diff
@@ -235,6 +235,16 @@ type ValidationRule struct {
+	// Since MessageExpression is used as a failure message, it must evaluate to a string.
+	// If both Message and MessageExpression are present on a Rule, then MessageExpression will be used if validation
+	// fails. If MessageExpression results in a runtime error, the runtime error is logged, and the validation failure message is produced
+	// as if the MessageExpression field were unset.
```
we don't allow `message` to contain line-breaks. do we have the same restriction for values produced from messageExpression (either by forbidding them or by stripping them out)? if so, document and test for that

DangerOnTheRanger replied:
Right, `messageExpression` output is not allowed to have line breaks. This is tested, but you're right that it's not documented - I'll update the docs.

DangerOnTheRanger replied:
(marking as resolved, docs now mention that line breaks are not allowed in the output)

liggitt commented:
I didn't review the CEL bits, just the API type and validation change, leaning on @cici37 / @jpbetz for implementation review here.

API surface is small, just had a few questions/clarifications around edge cases that need docs/test coverage

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:
```diff
@@ -252,16 +255,102 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
 }
 
+const (
+	maxMessageExpressionLen = 5000
```
Note: Do we want to move this over to `k8s.io/apiserver/pkg/apis/cel/config.go` as well?

jpbetz replied:
nit: `5 * 1024` so this can be documented as a 5k limit?

DangerOnTheRanger replied:
`5 * 1024` sounds good to me, as does moving it to `config.go`. I'll make both changes and mark this as resolved.

DangerOnTheRanger replied:
(marking as resolved, limit has been changed to 5 * 1024 and moved to `config.go`)

DangerOnTheRanger commented:
/retest

liggitt commented:
/approve
API bits (types, docs, and validation) lgtm
@cici37 @jpbetz have lgtm for impl

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types_jsonschema.go:
```diff
@@ -196,6 +196,18 @@ type ValidationRule struct {
+	// as if the messageExpression field were unset. If messageExpression evaluates to an empty string, a string with only spaces, or a string
+	// that contains line breaks, then the validation failure message will also be produced as if the messageExpression field were unset, and
+	// the fact that messageExpression produced an empty string/string with only spaces/string with line breaks will be logged.
+	// messageExpression has access to all the same variables as the rule; the only difference is the return type.
```
@jpbetz - did the env scoping get put in place so CRD CEL validation didn't automatically get access to the authorizer variable? is there anything that verifies validation rules and messageExpressions don't have that variable?

cici37 replied:
Yes. CRD validation rules go different path from ValidatingAdmissionPolicy hence it will not have access to authorizer related variables

cici37 replied:
We will make sure messageExpression not allowed to access to authorizer in here: https://github.com/kubernetes/kubernetes/pull/116397

jpbetz replied:
> Joe Betz - did the env scoping get put in place so CRD CEL validation didn't automatically get access to the authorizer variable? 

Yes, that's all in place.

> is there anything that verifies validation rules and messageExpressions don't have that variable?

Yes, when the secondary authz PR merged it included this test: https://github.com/kubernetes/kubernetes/blob/90c3232de77aa0dd09b948ffdd27c575688fba8a/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation_test.go#L1762

cici37 commented:
/lgtm
/approve
Thank you!
